File: raytrace/Object/Albedo/Texture.cpp
#include "Object/Albedo/Texture.hpp"

#include <fstream>

#include <windows.h>

namespace Object {
	namespace Albedo {
		Texture::Texture(std::unique_ptr<Object::Texture<3>> texture)
			: mTexture(std::move(texture))
		{
			mTexture->generateMipMaps();
		}

		Object::Color Texture::color(const Math::Point2D &surfacePoint, const Math::Bivector2D &surfaceProjection) const
		{
			Object::Texture<3>::Value value = mTexture->sample(surfacePoint, surfaceProjection);

			return Object::Color(value.channels[0], value.channels[1], value.channels[2]);
		}
	}
}

File: src/object/albedo/texture.rs
use crate::geo;
use crate::object;

use geo::Bivec2;
use geo::Point2;

use object::Color;

pub struct Texture {
    texture: object::Texture<3>,
}

impl Texture {
    pub fn new(mut texture: object::Texture<3>) -> Texture {
        texture.generate_mipmaps();

        return Texture { texture };
    }
}

impl object::Albedo for Texture {
    fn color(&self, surface_point: Point2, surface_projection: Bivec2) -> Color {
        let values = self.texture.sample(surface_point, surface_projection);
        return Color::new(values[0], values[1], values[2]);
    }
}
